### PR TITLE
Fetch specified bill from NYC API

### PR DESCRIPTION
**Pull request by hancush**
Related to https://github.com/opencivicdata/python-legistar-scraper/issues/59

Allows user to pass in `matter_id` to update a specific bill, rather than having to update all the bills in relevant time period.

### CODE REVIEW

File: nyc/bills.py [hunk 1]
```diff
@@ -161,133 +161,150 @@ def sponsorships(self, matter_id):
...
-                    for vote in votes:
-                        raw_option = vote['VoteValueName'].lower()
 
-                        if raw_option == 'suspended':
-                            continue
+    def scrape(self, window=3, matter_id=None):
+        self.version_errors = []
+
+        if matter_id:
+            matters = [self.fetch(matter_id)]
```
Comment (fgregg) [inline nyc/bills.py]: I think `matter` would be a clearer name.

File: nyc/bills.py [hunk 2]
```diff
@@ -161,133 +161,150 @@ def sponsorships(self, matter_id):
...
-                    vote_event.add_source(legistar_web)
-                    vote_event.add_source(legistar_api + '/histories')
+            return vote_event
 
-                    for vote in votes:
-                        raw_option = vote['VoteValueName'].lower()
 
-                        if raw_option == 'suspended':
-                            continue
+    def scrape(self, window=3, matter_id=None):
```
Comment (fgregg) [inline nyc/bills.py]: Do we want to be able to pass in more than one matter_id in at a time?

File: nyc/bills.py [hunk 3]
```diff
@@ -161,133 +161,150 @@ def sponsorships(self, matter_id):
             yield sponsorship
 
 
-    def scrape(self, window=3):
-        if window:
-            n_days_ago = datetime.datetime.utcnow() - datetime.timedelta(float(window))
-        else:
-            n_days_ago = None
+    def format_matter(self, matter):
```
Comment (fgregg) [inline nyc/bills.py]: format_matter isn't quite right is it.... maybe `get_bill`?

File: nyc/bills.py [hunk 4]
```diff
@@ -161,133 +161,150 @@ def sponsorships(self, matter_id):
...
+        matter_id = matter['MatterId']
+        if matter_id in DUPLICATED_ACTIONS:
+            return
```
Comment (fgregg) [inline nyc/bills.py]: let's explicitly return a `None` here and everywhere else

File: nyc/bills.py [hunk 5]
```diff
@@ -161,133 +161,150 @@ def sponsorships(self, matter_id):
...
-                                          legislative_session=related_bill_session,
-                                          relation_type='companion')
+        return bill
 
-            try:
-                text = self.text(matter_id)
-            except KeyError:
-                version_errors.append(legistar_web)
-                continue
+    def format_event(self, bill, action):
```
Comment (fgregg) [inline nyc/bills.py]: +1 on breaking this into multiple methods. Let's find a better name than `format_event`

File: nyc/bills.py (hunk 2 above)
Comment (hancush) [reply nyc/bills.py]: That's a good question. To me, `fetch` implies you're getting one thing. Does it have a similar connotation to you?

Comment (hancush) [reply nyc/bills.py]: Ah, scratch that, I guess there's nothing stopping us from calling `fetch` multiple times downstream, eh?

Comment (hancush) [reply nyc/bills.py]: Bearing that in mind, it'd be great to support more than one ID. Good idea.

File: nyc/bills.py (hunk 1 above)
Comment (hancush) [reply nyc/bills.py]: Is `matters` ok given [this change](https://github.com/opencivicdata/scrapers-us-municipal/pull/182/commits/c52cca469f2babcd5509068645ea5fc6d5b539cd#diff-0d0f0450f469b3ad1598867cde71cb7bR291)?

File: nyc/bills.py [hunk 6]
```diff
@@ -161,133 +161,155 @@ def sponsorships(self, matter_id):
...
-                                          legislative_session=related_bill_session,
-                                          relation_type='companion')
+        return bill
 
-            try:
-                text = self.text(matter_id)
-            except KeyError:
-                version_errors.append(legistar_web)
-                continue
+    def get_vote_event(self, bill, action):
```
Comment (fgregg) [inline nyc/bills.py]: This doesn't seem like the quite right separation.

when you do 

```python
        act = bill.add_action(action['action_description'],
                              action['action_date'],
                              organization={'name': action['responsible_org']},
                              classification=action['classification'])
```

does the `act` object have a handle on the bill? if so, could you just pass in an `act`?

Comment (hancush) [reply nyc/bills.py]: We don't have a handle on the vote events there, but I agree the separation is a bit confusing. How about c9d9f37?

File: nyc/bills.py [hunk 7]
```diff
@@ -161,133 +161,156 @@ def sponsorships(self, matter_id):
...
+            for vote in votes:
+                raw_option = vote['VoteValueName'].lower()
 
-            for action, vote in self.actions(matter_id):
-                act = bill.add_action(action['action_description'],
-                                      action['action_date'],
-                                      organization={'name': action['responsible_org']},
-                                      classification=action['classification'])
+                if raw_option == 'suspended':
+                    return None
```
Comment (fgregg) [inline nyc/bills.py]: this should be a `continue` I'm pretty sure.

File: nyc/bills.py [hunk 8]
```diff
@@ -161,133 +161,156 @@ def sponsorships(self, matter_id):
...
-                text = self.text(matter_id)
-            except KeyError:
-                version_errors.append(legistar_web)
-                continue
 
+    def get_vote_event(self, bill, act, vote):
+        '''Make VoteEvent object from given Bill, action, and vote.'''
+        result, votes = vote
+
+        if result:
```
Comment (fgregg) [inline nyc/bills.py]: it think I would be this check outside the method and just pass in votes.

File: nyc/bills.py (hunk 7 above)
Comment (hancush) [reply nyc/bills.py]: great catch!

**Pull request merged**